qiskit/ignis/measurement/discriminator/discriminators.py

```diff
@@ -0,0 +1,426 @@
+
+        self._expected_state = {}
+        for idx, schedule in enumerate(schedules):
+            if isinstance(schedule, Schedule):
+                name = schedule.name
+            else:
+                name = schedule
+            expected_state = expected_states[idx]
+            self._expected_state[name] = expected_state
+
+        # Used to rescale the xdata qubit by qubit.
+        self._description = None
+        self._standardize = standardize
+        self._scaler = None
+        self._qubit_mask = qubit_mask
+        self._schedules = schedules
+        self._backend_result_list = []
+        self._fitted = False
+
+        if cal_results is not None:
+            if isinstance(cal_results, list):
+                for result in cal_results:
+                    self._backend_result_list.append(result)
+            else:
+                self._backend_result_list.append(cal_results)
+
+        self._xdata = self.get_xdata(self._backend_result_list, schedules)
+        self._ydata = self.get_ydata(self._backend_result_list, schedules)
+
+    def _add_ydata(self, schedule: Union[Schedule, str]):
+        """
+        Adds the expected state of schedule to self._ydata.
+        Args:
+            schedule: schedule or schedule name.
+            Used to get the expected state.
+        """
+        if isinstance(schedule, Schedule):
+            self._ydata.append(self._expected_state[schedule.name])
+        else:
+            self._ydata.append(self._expected_state[schedule])
+
+    def add_data(self, result, refit=True, expected_state=None):
+        """
+        Args:
+            result: the Result obtained from e.g. backend.run().result()
+            recalc: this parameter is irrelevant and only needed for Liskov
+            principle.
+            refit: refit the discriminator if True.
+            expected_state: the expected state of the discriminator.
+        """
+        pass
+
+    @property
+    def expected_states(self):
+        return self._expected_state
+
+    @property
+    def fitted(self):
+        """True if the discriminator has been fitted to calibration data."""
+        return self._fitted
+
+    @abstractmethod
+    def _scale_data(self, xdata: List[List[float]], refit=False):
+        pass
+
+    @abstractmethod
+    def get_xdata(self, results: Union[Result, List[Result]],
+                  schedules: Union[List[str], List[Schedule]] = None) \
+            -> List[List[float]]:
+        """
+        Retrieves feature data (xdata) for the discriminator.
+        Args:
+            results: the get_memory() method is used to retrieve the level 1
+                data. If result is a list of Result then the first Result to
+                return the data of schedule in schedules is used.
+            schedules: Either the names of the schedules or the schedules
+                themselves.
+        Returns:
+            x data as a list of lists.
+        """
+        pass
```

Comment:
Perhaps standard implementations of this and `get_ydata` should be provided, or are you trying to handle the case of discriminating based on measurement level 0 data? If so I think this is fine.

Reply:
I wanted to leave the base discriminator independent of the data. Different backends and technologies might handle the level 1 data differently. E.g. Ions vs SC qubits: the format of the `nd.array` from `get_data` will be the same but there might be an extra data processing and formatting step between the `get_memory(...)` and the discriminator's `xdata` that is dependent on the technology/backend.